File: javascript/challenges/dynamic-array.js
const QUERY_TYPE_1 = 1;

const buildSeqList = n => {
  const list = [];
  for (let i = 0; i < n; i++) {
    list.push([]);
  }
  return list;
};

function dynamicArray(n, queries) {
  const sequences = buildSeqList(n);
  let lastAnswer = 0;

  const results = [];

  queries.forEach(query => {
    const [type, x, y] = query;
    const index = (x ^ lastAnswer) % n;
    const sequence = sequences[index];
    if (type === QUERY_TYPE_1) {
      sequence.push(y);
    } else {
      lastAnswer = sequence[y % sequence.length];
      results.push(lastAnswer);
    }
  });

  return results;
}

File: java/src/main/java/hackerrank/john/col/arrays/DynamicArray.java
package hackerrank.john.col.arrays;

public class DynamicArray {

  static List<List<Integer>> buildSeqList(int n) {
    List<List<Integer>> list = new ArrayList<>();
    for (int i = 0; i < n; i++) {
      List<Integer> inner = new ArrayList<>();
      list.add(inner);
    }
    return list;
  }

  public static List<Integer> dynamicArray(int n, List<List<Integer>> queries) {
    List<List<Integer>> sequences = buildSeqList(n);
    int lastAnswer = 0;

    List<Integer> results = new ArrayList<>();
    for (List<Integer> query: queries) {
      int type = query.get(0);
      int x = query.get(1);
      int y = query.get(2);

      int index = (x ^ lastAnswer) % n;
      List<Integer> sequence = sequences.get(index);

      if (type == 1) {
        sequence.add(y);
      } else {
        lastAnswer = sequence.get(y % sequence.size());
        results.add(lastAnswer);
      }
    }

    return results;
  }

}
